### PR TITLE
NFT Lock Mechanism

### PR DESCRIPTION
**Pull request by HashWarlock**
## Targets
* [x] Temporarily add Uniques pallet
* [x] Add a dependency injection Trait to Uniques pallet for RMRK Core pallet to implement
* [x] Implement Lock logic in RMRK Core
* [x] Add tests to ensure Locks operate correctly
* [x] Rebase changes with main branch
* [x] Submit PR for Uniques pallet in Substrate repo to add the dependency injection Trait
* [x] After PR accepted remove Uniques pallet from RMRK repo
* [x] Make PR suggested changes
* [x] Update Cargo to `polkadot-v0.9.22` when it is released with Locker Trait

## Procedure
* Temporarily add a local copy of the Uniques pallet to the repo
* Implement the dependency injection Trait for the `Lock` and implement a function to default to `false`
* Implement the new `Lock` Trait in RMRK Core and add logic to configure the `Lock` status as a `bool`
* Add error checks and tests to ensure a NFT has the following limitations:
  * no adding children 
  * no changing priority 
  * no adding resources 
  * no equip/unequip
  * no mutating attributes or properties
* Psuedo code ex: https://gist.github.com/h4x3rotab/152d05c9b5a18462e8dfe2b299b58db8

## RMRK Spec
- [LIST](https://github.com/rmrk-team/rmrk-spec/blob/master/standards/rmrk2.0.0/interactions/list.md)

### CODE REVIEW

**Comment by HashWarlock** [comment]
[Substrate PR 11025](https://github.com/paritytech/substrate/pull/11025)
[Cumulus PR 1085](https://github.com/paritytech/cumulus/pull/1085)

File: traits/Cargo.toml
Comment (ilionic) [inline traits/Cargo.toml:29]: can we switch back to `paritytech` repo yet?

**Review by ilionic — CHANGES_REQUESTED** [review]
Great work, would be good to merge it to avoid conflict. One question before we can merge.

File: traits/Cargo.toml
Comment (HashWarlock) [reply traits/Cargo.toml:29]: Hasn't been merged, yet. We could create a substrate repo for `rmrk-team` and put the fix there until the code is merged to `substrate`. I have the 2 PRs here https://github.com/paritytech/substrate/pull/11025
https://github.com/paritytech/cumulus/pull/1085 

I'm gonna reach out on the Matrix server to see if we can get another review.

**Review by bmacer — APPROVED** [review]
LGTM

**Comment by bmacer** [comment]
looks like your parity PRs are slowly making progress.  do we want to merge this before the switch back to parity is possible?  getting a little scared of the conflicts we'll be creating with getting too deep in multiple branches.

**Comment by HashWarlock** [comment]
> looks like your parity PRs are slowly making progress. do we want to merge this before the switch back to parity is possible? getting a little scared of the conflicts we'll be creating with getting too deep in multiple branches.

I will update the packages and rebase once it is put into the release. I don't think the changes will be too hard.

**Comment by HashWarlock** [comment]
Not currently included in `polkadot-v0.9.20` will wait until included in release. Code is rebased. I'll also be upgrading the toolchain whenever I merge.

**Comment by HashWarlock** [comment]
@ilionic All checks pass. We should merge this next given the amount of changes & upgrades to package dependencies.

**Comment by HashWarlock** [comment]
Officially released. We can merge now

**Review by ilionic — APPROVED** [review]

**Pull request merged**
